### PR TITLE
OpenMP reproducibility

### PR DESCRIPTION
**Pull request by Mikolaj-A-Kowalski**
First of all I think this one still requires a bit of work and more testing. I verified that things work on my laptop, but I haven't properly studied the impact on scaling. 

Basically this PR makes the parallel calculations with OpenMP reproducible. The following main changes were required: 
- `source_inter` sampling procedure had to be modified to skip from the initial RNG state (not the current thread-private). I'm afraid this bug avoided death in PR before...
- The incorrect, thread-private, RNG was used for sampling of the renormalisation of the source 
- The `particleDungeon` sorting has been implemented based on the `showerID` which is basically the sequence in this cycle dungeon of the particle. I am not happy with the name but I couldn't figure out a better one! (Also addresses #52) 
  - The sorting is done with some extra memory allocated at each cycle to store the permutation and count arrays. 
  - The permutation is done in place to conserve memory requirements (but it makes parallelisation not possible :-/, I am not sure it was a correct decision) 
  - Generally the sorting and renormalisation of the `particleDungeon` are done in serial and I fear a bit about scalability as a result of it.    
  
  P.S. Also I realised now that I forgot to add extra tests :-/ Treat it like a draft for now, please.

### CODE REVIEW

**Comment by ChasingNeutrons** [comment]
I must say showerID is quite a baffling name to me. What's the logic for choosing this name? This is just a unique ID for a given particle? Or if not, could you explain what it is supposed to do? Maybe something like queueID? Or threadID?

**Comment by Mikolaj-A-Kowalski** [comment]
> I must say showerID is quite a baffling name to me. What's the logic for choosing this name? This is just a unique ID for a given particle? Or if not, could you explain what it is supposed to do? Maybe something like queueID? Or threadID?

So the name comes from my photon warped brain and the 'EM shower' term. Basically to denote a source particle and all its progeny (with fission assumed to be absorption).  It is not a uniqueID of a particle per se as it is can be shared by multiple particles and their values are reset each cycle. `queueID` may be better but still makes me think of a identity of some queue, not the correlated bunch of particles. 

Basically name is like it is because I couldn't think of anything good and 'shower' was the only thing I could think of and I didn't want to stall the development by not being able to think of a name. But I am very open for suggestions. I did presume that `showerID` will be a placeholder really.

**Comment by ChasingNeutrons** [comment]
I feared it was something perverse like this! geneticID? familyID? spawnID (didn't that feature somewhere else?)?

**Comment by ChasingNeutrons** [comment]
Also parentID? Maybe the most important thing is to have a comment in particle explaining what it is.

**Comment by Mikolaj-A-Kowalski** [comment]
> I feared it was something perverse like this! geneticID? familyID? spawnID (didn't that feature somewhere else?)?

Perverse!!! 🤕... :cry: Auch! 
The problem is that I am not happy with any of them it is more like: `sequence of the parent particle in the Dungeon/source` but that does not exactly make a good variable name...

**Comment by ChasingNeutrons** [comment]
It's certainly hard to make that compact. If it's just to denote a group of particle together (and we don't want to use the word 'group'!) then what about cohortID?

**Comment by Mikolaj-A-Kowalski** [comment]
> It's certainly hard to make that compact. If it's just to denote a group of particle together (and we don't want to use the word 'group'!) then what about cohortID?

I am happy with the `cohortID`, I don't have better ideas to be honest. Unless @valeriaRaffuzzi has something in mind? 
Otherwise I will change `showerID` to `cohortID`

**Comment by valeriaRaffuzzi** [comment]
It is like a familyID but just for one generation, right? broodID? 

Google tells me it's a collective name for a bunch of siblings.

File: ParticleObjects/particleDungeon_class.f90 [hunk 1]
```diff
@@ -541,10 +610,13 @@ subroutine printToFile(self, name)
 
     ! Print out each particle co-ordinate
     do i = 1, self % pop
-      write(10,'(8A)') numToChar(self % prisoners(i) % r), &
-                       numToChar(self % prisoners(i) % dir), &
-                       numToChar(self % prisoners(i) % E), &
-                       numToChar(self % prisoners(i) % G)
+      ! write(10,'(8A)') numToChar(self % prisoners(i) % r), &
```
Comment (valeriaRaffuzzi) [inline ParticleObjects/particleDungeon_class.f90:613]: Should we remove the commented bit here?

File: ParticleObjects/particle_class.f90 [hunk 2]
```diff
@@ -649,31 +654,6 @@ function equal_particleState(LHS,RHS) result(isEqual)
     end if
   end function equal_particleState
 
-!  !!
```
Comment (valeriaRaffuzzi) [inline ParticleObjects/particle_class.f90:652]: Love a good clean up!

File: PhysicsPackages/eigenPhysicsPackage_class.f90 [hunk 3]
```diff
@@ -195,6 +195,8 @@ subroutine cycles(self, tally, tallyAtch, N_cycles)
         ! Obtain particle current cycle dungeon
         call self % thisCycle % copy(neutron, n)
 
+        neutron % showerID = n
```
Comment (valeriaRaffuzzi) [inline PhysicsPackages/eigenPhysicsPackage_class.f90:198]: Could this line be included into `particleDungeon % copy` instead of here? I think it would look neater.

File: PhysicsPackages/fixedSourcePhysicsPackage_class.f90 [hunk 4]
```diff
@@ -181,6 +181,7 @@ subroutine cycles(self, tally, N_cycles)
 
         ! Obtain particle from dungeon
         call self % thisCycle % copy(p, n)
+        p % showerID = n
```
Comment (valeriaRaffuzzi) [inline PhysicsPackages/fixedSourcePhysicsPackage_class.f90:184]: Same comment as before about the `copy`.

Comment (Mikolaj-A-Kowalski) [reply PhysicsPackages/fixedSourcePhysicsPackage_class.f90]: Applied

File: PhysicsPackages/eigenPhysicsPackage_class.f90 (hunk 3 above)
Comment (Mikolaj-A-Kowalski) [reply PhysicsPackages/eigenPhysicsPackage_class.f90]: Applied

File: ParticleObjects/particleDungeon_class.f90 (hunk 1 above)
Comment (Mikolaj-A-Kowalski) [reply ParticleObjects/particleDungeon_class.f90]: Has been removed now

**Comment by Mikolaj-A-Kowalski** [comment]
I think that all the comments have been addressed. 
Unrelated to this PR I added a commit to fix an issue with reuse of random numbers after source generation. it is not clean to put it here but was afraid I would forget otherwise ;-) 

I have also updated the documentation to highlight that the common buffer in fixed source calculation will (most likely) destroy OpenMP reproducibility.

**Comment by Mikolaj-A-Kowalski** [comment]
> I think that all the comments have been addressed. Unrelated to this PR I added a commit to fix an issue with reuse of random numbers after source generation. it is not clean to put it here but was afraid I would forget otherwise ;-)

I think that when I did this change i did not think sufficiently about fixed source calculations. I think there would be a reuse then still.

**Comment by Mikolaj-A-Kowalski** [comment]
> > I think that all the comments have been addressed. Unrelated to this PR I added a commit to fix an issue with reuse of random numbers after source generation. it is not clean to put it here but was afraid I would forget otherwise ;-)
> 
> I think that when I did this change i did not think sufficiently about fixed source calculations. I think there would be a reuse then still.

The commit regarding the reuse of random numbers was dropped.

**Review by valeriaRaffuzzi — APPROVED** [review]
It all looks good to me now! Thanks for working on this :)

**Review by ChasingNeutrons — APPROVED** [review]
One typo and then I can merge it in!

File: ParticleObjects/particle_class.f90 [hunk 5]
```diff
@@ -111,6 +103,7 @@ module particle_class
     integer(shortInt)          :: fate = 0       ! Neutron's fate after being subjected to an operator
     integer(shortInt)          :: type           ! Particle type
     integer(shortInt)          :: collisionN = 0 ! Index of the number of collisions the particle went through
+    integer(shortInt)          :: broodID = 0    ! ID of the shower (source particle number)
```
Comment (valeriaRaffuzzi) [inline ParticleObjects/particle_class.f90:106]:
```suggestion
    integer(shortInt)          :: broodID = 0    ! ID of the brood (source particle number)
```

File: ParticleObjects/Tests/particleDungeon_test.f90 [hunk 6]
```diff
@@ -271,5 +273,75 @@ subroutine testNormPopUp()
 
   end subroutine testNormPopUp
 
+  !!
+  !! Test sorting of the population by shower ID without duplicates
```
Comment (valeriaRaffuzzi) [inline ParticleObjects/Tests/particleDungeon_test.f90:277]:
```suggestion
  !! Test sorting of the population by brood ID without duplicates
```

File: ParticleObjects/Tests/particleDungeon_test.f90 [hunk 7]
```diff
@@ -271,5 +273,75 @@ subroutine testNormPopUp()
...
+    ! Initialise
+    call dungeon % init(10)
+
+    ! Store some particles with shower ID in reverse order
+    do i = 1,10
+      p % broodID = 10 - i + 1
+      call dungeon % detain(p)
+    end do
+
+    ! Sort by shower ID
```
Comment (valeriaRaffuzzi) [inline ParticleObjects/Tests/particleDungeon_test.f90:295]:
```suggestion
    ! Store some particles with brood ID in reverse order
    do i = 1,10
      p % broodID = 10 - i + 1
      call dungeon % detain(p)
    end do

    ! Sort by brood ID
```

File: ParticleObjects/Tests/particleDungeon_test.f90 [hunk 8]
```diff
@@ -271,5 +273,75 @@ subroutine testNormPopUp()
...
+    ! that the insertion order is preserved (for particles with the same shower ID)
+    do j = 1,N_duplicates
+      do i = 1,10
+        p % broodID = 10 - i + 1
+        p % G = j
+        call dungeon % detain(p)
+      end do
+    end do
+
+    ! Sort by shower ID
```
Comment (valeriaRaffuzzi) [inline ParticleObjects/Tests/particleDungeon_test.f90:332]:
```suggestion
  !! Test sorting of the population by brood ID with duplicates
  !!
  @Test
  subroutine testSortingByBroodID_withDuplicates()
    type(particleDungeon)    :: dungeon
    type(particleState)      :: p
    integer(shortInt)        :: i, j
    integer(shortInt), parameter :: N_duplicates = 7
    real(defReal), parameter :: TOL = 1.0E-9

    ! Initialise
    call dungeon % init(10 * N_duplicates)

    ! Store some particles with brood ID in reverse order
    ! Use the group number to distinguish duplicates and make sure
    ! that the insertion order is preserved (for particles with the same brood ID)
    do j = 1, N_duplicates
      do i = 1, 10
        p % broodID = 10 - i + 1
        p % G = j
        call dungeon % detain(p)
      end do
    end do

    ! Sort by brood ID
```

File: ParticleObjects/particleDungeon_class.f90 [hunk 9]
```diff
@@ -413,6 +424,10 @@ subroutine normSize(self, N, rand)
       call fatalError(Here,'Requested size: '//numToChar(N) //' is not +ve')
     end if
 
+    ! Determine the maximum shower ID and sort the dungeon
```
Comment (valeriaRaffuzzi) [inline ParticleObjects/particleDungeon_class.f90:427]:
```suggestion
    ! Determine the maximum brood ID and sort the dungeon
```

File: ParticleObjects/particleDungeon_class.f90 [hunk 10]
```diff
@@ -456,6 +471,69 @@ subroutine normSize(self, N, rand)
...
+    integer(shortInt), dimension(:), allocatable :: perm
+    type(particleState)                          :: tmp
+    character(100), parameter :: Here = 'sortBybroodID (particleDungeon_class.f90)'
+
+    ! Count number of particles with each shower ID
+    count = 0
+    do i = 1, self % pop
+      id = self % prisoners(i) % broodID
+
+      if (id < 1 .or. id > k) call fatalError(Here, 'Shower ID out of range: '//numToChar(id))
```
Comment (valeriaRaffuzzi) [inline ParticleObjects/particleDungeon_class.f90:494]:
```suggestion
  !! Reorder the dungeon so the brood ID is in the ascending order
  !!
  !! Args:
  !!   k [in] -> Maximum brood ID
  !!
  subroutine sortByBroodID(self, k)
    class(particleDungeon), intent(inout)        :: self
    integer(shortInt), intent(in)                :: k
    integer(shortInt), dimension(k)              :: count
    integer(shortInt)                            :: i, id, loc, c
    integer(shortInt), dimension(:), allocatable :: perm
    type(particleState)                          :: tmp
    character(100), parameter :: Here = 'sortBybroodID (particleDungeon_class.f90)'

    ! Count number of particles with each brood ID
    count = 0
    do i = 1, self % pop
      id = self % prisoners(i) % broodID

      if (id < 1 .or. id > k) call fatalError(Here, 'Brood ID out of range: '//numToChar(id))
```

**Comment by Mikolaj-A-Kowalski** [comment]
Thanks for catching the showeIds in the comments!

**Comment by valeriaRaffuzzi** [comment]
> Thanks for catching the showeIds in the comments!

No problem, it was a jetlag late night activity! Feel free to merge this when the comments are fixed :)

**Comment by Mikolaj-A-Kowalski** [comment]
Grep says it is fixed now. No 'shower' in SCONE. Merging
